kernel/src/hil/digest.rs

```diff
@@ -0,0 +1,62 @@
+//! Interface for Digest
+
+use crate::returncode::ReturnCode;
+
+/// Implement this trait and use `set_client()` in order to receive callbacks.
+pub trait Client<'a> {
+    /// This callback is called when the data has been added to the digest
+    /// engine.
+    /// On error `result` will contain a return code
+    /// On error or success `digest` will contain a reference to the original
+    /// data supplied to `add_data()`.
+    fn add_data_done(&'a self, result: Result<(), ReturnCode>, data: &'static mut [u8]);
+
+    /// This callback is called when a digest is computed.
+    /// On error `result` will contain device specific error codes
+    /// On error or success `digest` will contain a reference to the original
+    /// digest supplied to `run()`.
+    fn hash_done(&'a self, result: Result<(), u32>, digest: &'static mut [u8]);
+}
+
+pub trait DigestType: Eq + Copy + Clone + Sized + AsRef<[u8]> + AsMut<[u8]> {}
+
+impl DigestType for [u8; 32] {}
+
+/// Computes a digest (cryptographic hash) over data
+pub trait Digest<'a, T: DigestType> {
+    /// Set the client instance which will receive `hash_done()` and
+    /// `add_data_done()` callbacks.
+    /// This callback is called when the data has been added to the digest
+    /// engine.
+    /// The callback should follow the `Client` `add_data_done` callback.
+    fn set_client(&'a self, client: &'a dyn Client<'a>);
+
+    /// Add data to the digest block. This is the data that will be used
+    /// for the hash function.
+    /// Returns the number of bytes parsed on success
+    /// There is no guarentee the data has been written until the `add_data_done()`
+    /// callback is fired.
+    /// On error the return value will contain a return code and the original data digest
+    fn add_data(&self, data: &'static mut [u8]) -> Result<usize, (ReturnCode, &'static mut [u8])>;
+
+    /// Request the hardware block to generate a Digest and stores the returned
+    /// digest in the memory location specified.
+    /// This doesn't return any data, instead the client needs to have
+    /// set a `hash_done` handler to determine when this is complete.
+    /// On error the return value will contain a return code and the original digest
+    /// If there is data from the `add_data()` command asyncrously waiting to
+    /// be written it will be written before the operation starts.
+    fn run(&'a self, digest: &'static mut T) -> Result<(), (ReturnCode, &'static mut T)>;
```

Comment:
I'm still curious what should happen if a user calls `run` without ever calling `set_mode_hmacsha256()`.

Reply:
Right now it would just work I think as it's the only option.

Reply:
What happens if there are two options?

I think conceptually I like the `Digest<DigestType>` approach from #1727 more, primarily because this question doesn't exist, and there is no confusion about what it means to directly use a `Digest` interface. However, practically I think the HmacAlgorithm::set_mode() approach in this PR is more flexible and usable and leads to easier to read code, particularly with a single hardware peripheral that can support numerous modes.

But, given this is a HIL, it should describe what an implementer is supposed to do if a user uses it incorrectly (i.e. doesn't call `set_mode_X()`). Should it just pick one? Does it return an error (which error)? Does it panic?

Reply:
I added some documentation describing what should be done.